include/xtensor/xsemantic.hpp

```diff
@@ -18,8 +18,211 @@
 
 namespace xt
 {
+    /* Forward declarations */
+
+    template <class D>
+    class xview_semantic;
+
+    template <class CT, class X>
+    class xbroadcast;
+
+    template <class F, class R, class S>
+    class xgenerator;
+
     namespace detail
     {
+
+        template <class T, class Enable = void>
+        struct has_memory_address : std::false_type
+        {
+        };
+
+        template <class T>
+        struct has_memory_address<T, void_t<decltype(std::addressof(*std::declval<T>().begin()))>> : std::true_type
+        {
+        };
+
+        struct memory_range
+        {
+            // Checking pointer overlap is more correct in integer values,
+            // for more explanation check https://devblogs.microsoft.com/oldnewthing/20170927-00/?p=97095
+            const uintptr_t m_first = 0;
+            const uintptr_t m_last = 0;
+
+            explicit memory_range() = default;
+
+            template <class T>
+            explicit memory_range(T* first, T* last)
+                : m_first(reinterpret_cast<uintptr_t>(last < first ? last : first)),
+                m_last(reinterpret_cast<uintptr_t>(last < first ? first : last)) {}
+
+            template <class T>
+            bool overlaps(T* first, T* last) const
+            {
+                if (first <= last)
+                {
+                    return reinterpret_cast<uintptr_t>(first) <= m_last && reinterpret_cast<uintptr_t>(last) >= m_first;
+                }
+                else
+                {
+                    return reinterpret_cast<uintptr_t>(last) <= m_last && reinterpret_cast<uintptr_t>(first) >= m_first;
+                }
+            }
+        };
+
+        template <class E, class Enable = void>
+        struct overlapping_memory_checker_traits
+        {
+            static bool check_overlap(const E&, const memory_range&)
+            {
+                return true;
+            }
+        };
+
+        template <class E>
+        struct overlapping_memory_checker_traits<E, std::enable_if_t<has_memory_address<E>::value>>
+        {
+            static bool check_overlap(const E& expr, const memory_range& dst_range)
+            {
+                if (expr.size() == 0)
+                {
+                    return false;
+                }
+                else
+                {
+                    return dst_range.overlaps(std::addressof(*expr.begin()), std::addressof(*expr.rbegin()));
+                }
+            }
+        };
```

Comment:
All these definitions should probably go into `xutils.hpp` or `xassign.hpp`

Reply:
ok, I'll put it in xutils.hpp